R/tiq-test.R

```diff
@@ -0,0 +1,606 @@
+    ti.dt = loadTI("raw", group, date=str.date)
+    if (!is.null(ti.dt)) {
+      split.ti = split(ti.dt, ti.dt$source)
+
+      ## Performing the novelty test only on the sources we select
+      if (!is.null(select.sources)) {
+        split.names = intersect(select.sources, names(split.ti))
+      } else {
+        split.names = names(split.ti)
+      }
+
+      for (name in split.names) {
+        setkey(split.ti[[name]], entity)
+        split.ti[[name]] = unique(split.ti[[name]])
+
+        ti.count[[name]][[str.date]] = nrow(split.ti[[name]])
+        if (!is.null(prev.split.ti) && !is.null(prev.split.ti[[name]])) {
+          ti.added.ratio[[name]][[str.date]] = tiq.helper.differenceCount(split.ti[[name]], prev.split.ti[[name]]) /
+                                               ti.count[[name]][[str.date]]
+          ti.churn.ratio[[name]][[str.date]] = tiq.helper.differenceCount(prev.split.ti[[name]], split.ti[[name]]) /
+                                               ti.count[[name]][[str.date]]
+
+          ## This adjustment is necessary to compensate for data collection issues
+          ## we had on outbound feeds. You can't really change more then everything you got
+          if (ti.added.ratio[[name]][[str.date]] > 0.6) ti.added.ratio[[name]][[str.date]] = 0
+          if (ti.churn.ratio[[name]][[str.date]] > 1) ti.churn.ratio[[name]][[str.date]] = 0
+        }
+      }
+    } else {
+      split.ti = NULL
+    }
+    prev.split.ti = split.ti
+  }
+  if (.progress) close(pb)
+
+  return(list(ti.count=ti.count, ti.added.ratio=ti.added.ratio, ti.churn.ratio=ti.churn.ratio))
+}
+
+#' Plots the results of the novelty test in a (mostly) clear graph for comparisons
+#'
+#' @param novelty the output of the 'tiq.test.noveltyTest' function
+#' @param plot.sources a chararacter vector of the sources on the novelty test you want
+#'        to be a part of the plot, or NULL for all of them
+#' @export
+
+plotNoveltyTest <- function(novelty, sources=NULL) {
+
+  test_that("tiq.test.plotNoveltyTest: parameters must have correct types", {
+    expect_that(class(novelty), equals("list"))
+    expect_that(class(novelty$ti.count), equals("list"))
+    expect_that(class(novelty$ti.added.ratio), equals("list"))
+    expect_that(class(novelty$ti.churn.ratio), equals("list"))
+  })
+
+  if (is.null(sources)) {
+    plot.sources = names(novelty$ti.added.ratio)
+  }
+
+  tmp <- as.data.frame(novelty$ti.added.ratio)
+  tmp$date <- rownames(tmp)
+  rownames(tmp) <- NULL
+  added_ratio <- gather(tmp, source, added_ratio, -date)
+
+  tmp <- as.data.frame(novelty$ti.churn.ratio)
+  tmp$date <- rownames(tmp)
+  rownames(tmp) <- NULL
+  churn_ratio <- gather(tmp, source, churn_ratio, -date)
+
+  tmp <- as.data.frame(novelty$ti.count)
+  tmp$date <- rownames(tmp)
+  rownames(tmp) <- NULL
+  ti_count <- gather(tmp, source, ti_count, -date)
+
+  tmp <- merge(merge(added_ratio, churn_ratio), ti_count)
```

Comment:
This whole `tmp` segment fails if for some reason there is missing data on `novelty` list. 

Sadly, this is allowed on the current summarization implementation, and was handled correctly by the previous code. It will have ti be fixed either here or there.